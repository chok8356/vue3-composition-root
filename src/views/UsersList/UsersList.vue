<template>
  <div :class="$style.users">
    <div v-if="error">{{ error }}</div>
    <RouterLink
      v-for="user in users"
      :key="user.id"
      :class="$style.user"
      data-test="user"
      :to="{ name: 'UserDetail', params: { userId: user.id } }">
      {{ user.name }}
    </RouterLink>
  </div>
</template>

<script setup lang="ts">
import type { UsersListDeps } from '@/views/UsersList/UsersListDeps'

import { onMounted, ref } from 'vue'
import { RouterLink } from 'vue-router'

import type { IUser } from './types/IUser'

const props = defineProps<{
  deps: UsersListDeps
}>()

const users = ref<IUser[]>([])

const error = ref('')

onMounted(async () => {
  await props.deps.getUsers().then((data) => {
    data.match({
      err: async (x) => {
        error.value = x
      },
      ok: async (x) => {
        users.value = x
      },
    })
  })
})
</script>

<style module>
.users {
  display: flex;
  flex-direction: column;
  gap: 1rem;
  width: 100%;
}

.user {
  display: flex;
  width: 100%;
}
</style>
