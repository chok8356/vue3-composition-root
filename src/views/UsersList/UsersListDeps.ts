import type { IUser } from '@/views/UsersList/types/IUser'
import type { Result } from '@thames/monads'

import { fetchClient } from '@/api/fetchClient'
import { jsonPlaceholderApiClient } from '@/api/jsonplaceholder/jsonPlaceholderApiClient'
import { getUsersFactory } from '@/views/UsersList/deps/getUsersFactory'

export type UsersListDeps = {
  getUsers: () => Promise<Result<IUser[], string>>
}

export const usersListDeps: UsersListDeps = {
  getUsers: getUsersFactory(jsonPlaceholderApiClient(fetchClient).getUsers),
}
