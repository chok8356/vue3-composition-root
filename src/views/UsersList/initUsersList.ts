import type { App } from 'vue'

import {
  type GetUsersDelegate,
  getUsersInjectionKey,
} from '@/views/UsersList/ports/GetUsersDelegate'

export interface Ports {
  getUsers: GetUsersDelegate
}

export const initUsersList = (app: App, { getUsers }: Ports) => {
  app.provide(getUsersInjectionKey, getUsers)
}
