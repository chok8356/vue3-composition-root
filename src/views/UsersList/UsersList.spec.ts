import type { UsersListDeps } from '@/views/UsersList/UsersListDeps'

import { Ok } from '@thames/monads'
import { flushPromises, shallowMount } from '@vue/test-utils'
import { describe, expect, it, vi } from 'vitest'

import type { IUser } from './types/IUser'

import UsersList from './UsersList.vue'

const users: IUser[] = [
  { id: 1, name: 'John Doe' },
  { id: 2, name: 'Jane Doe' },
]

const deps: UsersListDeps = {
  getUsers: vi.fn().mockResolvedValue(Ok(users)),
}

describe('UsersList.vue', () => {
  it('fetches users list and displays them correctly', async () => {
    const wrapper = shallowMount(UsersList, {
      props: {
        deps,
      },
    })
    await flushPromises()

    expect(deps.getUsers).toHaveBeenCalled()

    const userElement = wrapper.findAll('[data-test="user"]')

    expect(userElement.length).toBe(2)
    expect(userElement[0].text()).toContain('John Doe')
    expect(userElement[1].text()).toContain('Jane Doe')

    expect(wrapper.html()).toMatchSnapshot()
  })
})
