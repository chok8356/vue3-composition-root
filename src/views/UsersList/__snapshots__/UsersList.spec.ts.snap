// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`usersList.vue > fetches users list and displays them correctly 1`] = `
"<div class="users">
  <!--v-if-->
  <router-link-stub to="[object Object]" replace="false" custom="false" ariacurrentvalue="page" class="user" data-test="user">John Doe</router-link-stub>
  <router-link-stub to="[object Object]" replace="false" custom="false" ariacurrentvalue="page" class="user" data-test="user">Jane Doe</router-link-stub>
</div>"
`;
